docs/design_spec.md

```diff
@@ -0,0 +1,94 @@
+belongs to the same network fabric (L2/L3 domain). This prevents traffic from ever being forwarded back
+out the same interface from which it originated (1). Additionally, this design increases network
+function processing efficiency, because the cache of the tables are limited to lookups for traffic that may ingress
+a single port. Debugging becomes easier because the only traffic in a particular pipeline only ever comes
+from a single port. However, one of the drawbacks to this design is that there is increased memory usage,
+as the number of instances of network functions scale by the number of ports in the system.
+
+Although each Port contains its own Port Pipeline, the core network functions portion of the pipeline
+remain the same for Ports that belong to the same network fabric. Therefore Nimbess abstracts this portion of
+the Port Pipelines into what is called a Meta Pipeline. This allows Nimbess to make core updates to only
+a single Meta Pipeline, which will then be used to create or update Port Pipelines. Here is an example of
+what the pipelines may look like for a single L2 network with 2 Ports:
+
+[![Nimbess Pipelines](./img/Nimbess_Pipelines.png)](./img/Nimbess_Pipelines.png)
+
+In the example above, the EgressPort for Port 1 is attached to every other Port Pipeline (in this case only 1)
+other than Port 1's Pipeline. The same is true for Port 2. This behavior prevents forwarding packets out the
+same interface traffic came in on. Each instance of the Meta Pipeline for the Port Pipeline uses a unique ID
+in order to distinguish these functions when they are rendered into different data planes. There may be many
+Meta Pipelines which belong to either different L2 networks (i.e. VLANs) or L3 networks (i.e. VRFs). A Port
+may only belong to a single Meta Pipeline.
+
+## Southbound Design
+
+Nimbess Agent uses a data plane agnostic southbound. Nimbess defines an abstract Driver type which any
+data plane driver may implement. The driver is configured in the Nimbess Agent config file and loaded at
+run time. A data plane driver is responsible for taking in Port Pipeline objects translating them into driver
+specific configuration, and then updating the Port Pipeline Modules with any related network configuration.
+This returned network configuration may vary per module, but as an example, updating a new Port object with
+the MAC address returned from the data plane when the Port is created.
+
+## Example Code Flow - CNI ADD Request
+
+Below is an example of walking through the pseudo code of a CNI ADD request:
+
+[![CNI CodeFlow](./img/code2flow_IvSpoQ.svg)](./img/code2flow_IvSpoQ.svg)
+
+A Nimbess Pipeline structure is used to hold all of the information needed to define a network pipeline. It includes
+information such as an array of all modules in the pipeline, the MetaKey for this pipeline, which port this
+pipeline belongs to, etc. While the pipeline holds the information describing which modules belong to the
```

Comment:
is a module a function?

Reply:
yeah Module is a generic Network Function